contracts/ERC20UpgradeableWithMultiplier.sol

```diff
@@ -0,0 +1,261 @@
+     * Defined in 1e18 precision.
+     *
+     */
+    uint256 private _multiplier;
+
+    /**
+     * @dev Emitted when `value` token shares are moved from one account (`from`) to
+     * another (`to`).
+     *
+     * Note that `value` may be zero.
+     */
+    event TransferShares(address indexed from, address indexed to, uint256 value);
+
+    /**
+     * @dev Emitted when multiplier value is updated
+     */
+    event MultiplierUpdated(uint256 value);
+
+    /**
+     * @dev Sets the values for {name} and {symbol}.
+     *
+     * The default value of {decimals} is 18. To select a different value for
+     * {decimals} you should overload it.
+     *
+     * All two of these values are immutable: they can only be set once during
+     * construction.
+     */
+    function __Multiplier_init() internal onlyInitializing {
+        __Multiplier_init_unchained();
+    }
+
+    function __Multiplier_init_unchained() internal onlyInitializing {
+        _multiplier = 1e18;
+    }
+    
+    /**
+     * @dev See {IERC20-totalSupply}.
+     */
+    function totalSupply() public view virtual override returns (uint256) {
+        return _totalShares * _multiplier / 1e18;
+    }
+
+    /**
+     * @dev Returns ratio of shares to underlying amount in 18 decimals precision
+     */
+    function multiplier() public view virtual returns (uint256) {
+        return _multiplier;
+    }
+
+    /**
+     * @dev See {IERC20-balanceOf}.
+     */
+    function balanceOf(address account) public view virtual override returns (uint256) {
+        return _shares[account] * _multiplier / 1e18;
+    }
+
+    /**
+     * @dev Returns amount of shares owned by given account
+     */
+    function sharesOf(address account) public view virtual returns (uint256) {
+        return _shares[account];
+    }
+
+    /**
+     * @dev Transfers underlying shares to destination account
+     *
+     * Requirements:
+     *
+     * - `to` cannot be the zero address.
+     * - the caller must have a balance of at least `sharesAmount`.
+     */
+    function transferShares(address to, uint256 sharesAmount) public virtual returns (bool) {
+        address owner = _msgSender();
+        _transferShares(owner, to, sharesAmount);
+        return true;
+    }
+
+    /**
+     * @return the amount of shares that corresponds to `_underlyingAmount` underlying amount.
+     */
+    function getSharesByUnderlyingAmount(uint256 _underlyingAmount) public view returns (uint256) {
+        return _underlyingAmount
+             * 1e18
+             / _multiplier;
+    }
+
+    /**
+     * @return the amount of underlying that corresponds to `_sharesAmount` token shares.
+     */
+    function getUnderlyingAmountByShares(uint256 _sharesAmount) public view returns (uint256) {
+        return _sharesAmount
+             * _multiplier
+             / 1e18;
+    }
+
+    /**
+     * @notice Moves `_sharesAmount` shares from `from` to `to`.
+     *
+     * Requirements:
```

Comment:
let's keep comments alligned, if other functions have emits part it should be defined here as well